### PR TITLE
overlord: add infrastructure for simple state format/content migrations

### PR DESCRIPTION
**Pull request by pedronis**
This adds infrastructure for simple state format/content migrations.

For more complex migrations that may need online connections/retry it may be better to use actual Changes, though we might have to teach taskrunner/mustWait about conflicts to deal with in-flight changes that might interfere with the migration, might need to wait for migration success before proceeding.

Used by #1366, may be easier to review that for full context.

### CODE REVIEW

File: overlord/migrations.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+func runMigration(s *state.State, level int) error {
+	m := migrations[level]
+	if m == nil {
+		return fmt.Errorf("no supported migration")
+	}
+
+	commit := func() {
+		s.Set("patch-level", level+1)
+		s.Unlock()
```
Comment (niemeyer) [inline overlord/migrations.go]: I know what you're trying to do, but this is a bit surprising, confirmed by the fact we need documentation below detailing the unusual convention.

We could achieve the same goal with `bumpLevel()` instead which would hand just the level setting to the method itself to do, but I suggest we try something even simpler first, until we see it doesn't work: let's just wrap the whole migration on a lock ourselves, and do the level setting after it returns. We can defer the unlock.

File: overlord/migrations.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// 	s.Lock()
+// 	// get data to migrate from state...
+// 	s.Unlock()
+//
+// 	// prepare migrated data, can return error
+//
+// 	s.Lock()
+// 	// store back migrated data into state with s.Set etc, no error paths
+// 	commit()
+//}
```
Comment (niemeyer) [inline overlord/migrations.go]: With the simplification described, the docs above might become
```
// migrations maps from L to migration function for L to L+1

**Comment by niemeyer** [comment]
One point about the commit logic, and LGTM.

**Comment by pedronis** [comment]
retest this please

**Comment by mvo5** [comment]
:+1:

**Pull request merged**